src/devices/machine/cs8900a.cpp

```diff
@@ -0,0 +1,1564 @@
+  CS8900_PP_ADDR_CC_RXCTL   = 0x0104,      /* # RW - 4.4.8.,  p. 54 - 0005 */
+  CS8900_PP_ADDR_CC_TXCFG   = 0x0106,      /*   RW - 4.4.9.,  p. 55 - 0007 */
+  CS8900_PP_ADDR_CC_TXCMD   = 0x0108,      /*   R- - 4.4.11., p. 57 - 0009 */
+  CS8900_PP_ADDR_CC_BUFCFG  = 0x010A,      /*   RW - 4.4.12., p. 58 - 000B */
+  CS8900_PP_ADDR_CC_LINECTL = 0x0112,      /* # RW - 4.4.16., p. 62 - 0013 */
+  CS8900_PP_ADDR_CC_SELFCTL = 0x0114,      /*   RW - 4.4.18., p. 64 - 0015 */
+  CS8900_PP_ADDR_CC_BUSCTL  = 0x0116,      /*   RW - 4.4.20., p. 66 - 0017 */
+  CS8900_PP_ADDR_CC_TESTCTL = 0x0118,      /*   RW - 4.4.22., p. 68 - 0019 */
+  /* CS8900_PP_ADDR_STATUS_EVENT is subdivided: */
+  CS8900_PP_ADDR_SE_ISQ      = 0x0120,     /*   R- - 4.4.5.,  p. 51 - 0000 */
+  CS8900_PP_ADDR_SE_RXEVENT  = 0x0124,     /* # R- - 4.4.7.,  p. 53 - 0004 */
+  CS8900_PP_ADDR_SE_TXEVENT  = 0x0128,     /*   R- - 4.4.10., p. 56 - 0008 */
+  CS8900_PP_ADDR_SE_BUFEVENT = 0x012C,     /*   R- - 4.4.13., p. 59 - 000C */
+  CS8900_PP_ADDR_SE_RXMISS   = 0x0130,     /*   R- - 4.4.14., p. 60 - 0010 */
+  CS8900_PP_ADDR_SE_TXCOL    = 0x0132,     /*   R- - 4.4.15., p. 61 - 0012 */
+  CS8900_PP_ADDR_SE_LINEST   = 0x0134,     /*   R- - 4.4.17., p. 63 - 0014 */
+  CS8900_PP_ADDR_SE_SELFST   = 0x0136,     /*   R- - 4.4.19., p. 65 - 0016 */
+  CS8900_PP_ADDR_SE_BUSST    = 0x0138,     /* # R- - 4.4.21., p. 67 - 0018 */
+  CS8900_PP_ADDR_SE_TDR      = 0x013C      /*   R- - 4.4.23., p. 69 - 001C */
+};
+
+enum tx_rx_min_max_e : u16 {
+  MAX_TXLENGTH = 1518,
+  MIN_TXLENGTH = 4,
+  MAX_RXLENGTH = 1518,
+  MIN_RXLENGTH = 64
+};
+
+enum cs8900_tx_state_e : u8 {
+  CS8900_TX_IDLE       = 0,
+  CS8900_TX_GOT_CMD    = 1,
+  CS8900_TX_GOT_LEN    = 2,
+  CS8900_TX_READ_BUSST = 3
+};
+
+enum cs8900_rx_state_e : u8 {
+  CS8900_RX_IDLE      = 0,
+  CS8900_RX_GOT_FRAME = 1
+};
+
+enum pp_ptr_masks_e : u16 {
+  PP_PTR_AUTO_INCR_FLAG = 0x8000, /* auto increment flag in package pointer */
+  PP_PTR_FLAG_MASK      = 0xf000, /* is always : x y 1 1 (with x=auto incr) */
+  PP_PTR_ADDR_MASK      = 0x0fff  /* address portion of packet page pointer */
+};
+
+#define LO_BYTE(x) (u8)((x)& 0xff)
+#define HI_BYTE(x) (u8)(((x) >> 8)& 0xff)
+#define LOHI_WORD(x, y) ((u16)(x) | (((u16)(y)) << 8))
+
+// ------------------- END #defines ---------------------
+
+// ---- debug logging support ----
+char * cs8900a_device::debug_outbuffer(const int length, const unsigned char *const buffer)
+{
+  int i;
+  static const u32 MAXLEN_DEBUG             = 1600;
+  static const u32 TFE_DEBUG_MAX_FRAME_DUMP = 150;
+  static char outbuffer[MAXLEN_DEBUG * 4 + 1];
+  char *p = outbuffer;
+
+  assert(TFE_DEBUG_MAX_FRAME_DUMP <= MAXLEN_DEBUG);
+
+  *p = 0;
+
+  for (i = 0; i < TFE_DEBUG_MAX_FRAME_DUMP; i++)
+  {
+    if (i >= length)
+      break;
+
+    sprintf(p, "%02X%c", buffer[i], ((i + 1) % 16 == 0) ? '*' : (((i + 1) % 8 == 0) ? '-' : ' '));
+    p += 3;
+  }
+
+  return outbuffer;
+}
```

Comment:
Looking at this right now.  How do I actually turn my logging on now that it's grafted into the MAME logging system? `debug_outbuffer` is called only by a couple of specific logging statements, and the MAME docs aren't super clear on how to enable particular modules' logging.

Reply:
The -oslog switch at the command line turns on writing logs to the TTY, and I think there's some way to have it show in the debugger as well.

Reply:
I actually set `-log -oslog -verbose`, and I did see logs for the CPU but not for my chipset, even with `VERBOSE` set to `1`, which is what led me to ask in the first place.  I'm fairly sure I followed the recommended pattern for logging in my code.

Reply:
You need to set `VERBOSE` to the combination of log masks you want enabled, e.g.
```
#define VERBOSE (CS8900_DEBUG | CS8900_DEBUG_REGISTERS | CS8900_DEBUG_FRAMES)
```
If you set it to `1` it's equivalent to `LOG_GENERAL` and it will just enable stuff logged with the abbreviated `LOG` macro.